### PR TITLE
first version of metadata handling

### PR DESCRIPTION
**Pull request by jdroenner**
This is a first step to introduce metadata access: #20.

TODO: RasterBand metadata. This probably needs a `struct RasterBand`.

### CODE REVIEW

**Comment by frewsxcv** [comment]
> TODO: RasterBand metadata. This probably needs a struct RasterBand.

Are you planning to address this TODO in this PR or in a follow-up one? Either way is fine

**Comment by jdroenner** [comment]
I'm not yet sure how much work this will require. If you want to merge this early i can do a follow-up ;)

File: src/metadata.rs [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+extern {
+    fn GDALGetMetadataItem(hGdalMayorObject: *const c_void, pszName: *const c_char, pszDomain: *const c_char) -> *const c_char;
+    fn GDALSetMetadataItem(hGdalMayorObject: *const c_void, pszName: *const c_char, pszValue: *const c_char, pszDomain: *const c_char ) -> c_int;
+}
+
+pub trait Metadata: MajorObject {
+
+    fn get_metadata_item(&self, key: &str, domain: &str) -> Option<String> {
+        let c_key = CString::new(key.to_owned()).expect("Could not transform String to CString"); // FIXME: no unwrap
+        let c_domain = CString::new(domain.to_owned()).expect("Could not transform String to CString");
```
Comment (frewsxcv) [inline src/metadata.rs]: could these both just `return None` if the `CString::new` fails?

File: src/metadata.rs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        }
+        else {
+            Some(_string(c_res))
+        }
+    }
+
+    fn set_metadata_item(&mut self, key: &str, value: &str, domain: &str) -> Result<(), ()> {
+        let c_key = CString::new(key.to_owned()).expect("Could not transform String to CString"); // FIXME: no unwrap
+        let c_domain = CString::new(domain.to_owned()).expect("Could not transform String to CString");
+        let c_value =  CString::new(value.to_owned()).expect("Could not transform String to CString");
```
Comment (frewsxcv) [inline src/metadata.rs]: Same thing here. `return Err(())` instead of `expect`?

File: src/metadata.rs (hunk 1 above)
Comment (jdroenner) [inline src/metadata.rs]: Good catch. A `&str` which can not be converted to a `CString ` can't be a valid key.

File: src/metadata.rs (hunk 2 above)
Comment (jdroenner) [inline src/metadata.rs]: Yes you are right. I missed this after changing the method to use `&str` instead of `CStr`.

File: src/metadata.rs (hunk 1 above)
Comment (frewsxcv) [inline src/metadata.rs]: Need any help making this changes? I can make them for you if you want, just let me know

Comment (jdroenner) [inline src/metadata.rs]: I think i can do it till Friday :)

Comment (frewsxcv) [inline src/metadata.rs]: Sounds good, no rush!

**Comment by jdroenner** [comment]
I also added `GDALGetDescription ` and `impl Metadata for Driver`

**Comment by frewsxcv** [comment]
Looks great, thanks!

**Pull request merged**